src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -1302,6 +1302,7 @@ protected unsafe void WriteEventCore(int eventId, int eventDataCount, EventSourc
         [CLSCompliant(false)]
         protected unsafe void WriteEventWithRelatedActivityIdCore(int eventId, Guid* relatedActivityId, int eventDataCount, EventSource.EventData* data)
         {
+            m_activeWritesCount++;
```

Comment:
It occurred to me that the volatile solves the reader/writer ordering, there is still a race for readers updateing m_activeWritesCount. ++ is not atomic, so you could have a race where an increment or decrement gets lost. We should be using Interlocked.Increment and Interlocked.Decrement